crates/agent/src/publications.rs

```diff
@@ -196,9 +201,54 @@ impl PublishHandler {
             .await;
         }
 
-        let live_spec_ids: Vec<_> = spec_rows.iter().map(|row| row.live_spec_id).collect();
+        let expanded_rows = specs::expanded_specifications(row.user_id, &spec_rows, txn).await?;
+        tracing::debug!(specs = %expanded_rows.len(), "resolved expanded specifications");
+
+        let errors = draft::extend_catalog(
+            &mut draft_catalog,
+            expanded_rows
+                .iter()
+                .map(|r| (r.live_type, r.catalog_name.as_str(), r.live_spec.0.as_ref())),
+        );
+        if !errors.is_empty() {
+            anyhow::bail!("unexpected errors from expanded specs: {errors:?}");
+        }
+
+        // It's important that we prune unbound collection specs _after_ the catalog is expanded,
+        // and before checking resource quotas
+        let pruned_collections =
```

Comment:
I think we should be pruning in the DB, rather than via the `models::Catalog` and application code.

The reason why is that this change places new expectations on what, exactly, `expanded_specifications` is supposed to do which will make it difficult to refactor / update it's semantics. IMO `expanded_specifications` should be doing less, pulling fewer things into any given build. I think _probably_ it would all still be fine, but it also seems straightforward to identify specs to prune via a DB query instead.

The way this can work instead, is to first restore the previous control flow so that `apply_updates_for_row` is first called for each spec.

Next, immediately after `apply_updates_for_row`, you can query `live_specs` outer joined with `live_spec_flows`, and look for brand new rows having `spec_type = 'collection'` with no incoming or outgoing flows. That would give you the same result without any dependence on `expanded_specifications`.

ETA: OH! It's also not correct to call `expanded_specifications` _before_ applying updates for each row, because we need those updated `live_spec_flows` from `apply_updates_for_row` in order to power the expansion.